File: lib/commands/cowsay.rb
# frozen_string_literal: true

module Rebbot
  module Commands
    class Cowsay < Rebbot::Commands::Base
      on :cowsay, description: 'have a cow say something'

      def with_options(cmd)
        cmd.string('message', 'have a cow say something', required: true)
        cmd.string('type', 'type of cow')
      end

      def on_event(event)
        type = Cow.cows.find { |cow| cow == event.options['type'] } || 'default'
        msg = event.options['message'][0..Cow::MAX_LINE_LENGTH]
        event.respond(content: "```\n#{Cow.new(cow: type).say(msg)}```")
      end
    end
  end
end

File: src/commands/cowsay.ts
import { InteractionResponseType } from "discord-interactions";
import * as cowsay from "cowsay";
import { CommandOptionType, cmd } from "../types.js";

export default cmd(
  {
    name: "cowsay",
    description: "Have a cow say something",
    options: [
      {
        name: "message",
        description: "What the cow says",
        type: CommandOptionType.STRING,
        required: true,
      },
      {
        name: "type",
        description: "Cow type",
        type: CommandOptionType.STRING,
        required: false,
      },
    ],
  },
  async ({ message, type }) => {
    const opts: cowsay.IOptions = { text: message };
    if (type) opts.f = type;

    const art = cowsay.say(opts);
    return {
      type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
      data: { content: "```\n" + art + "\n```" },
    };
  },
);
